File: Codes/1652-defuse-the-bomb_20240505.py
'''
Author: LetMeFly
Date: 2024-05-05 14:03:07
LastEditors: LetMeFly
LastEditTime: 2024-05-05 14:07:35
'''
from typing import List

class Solution:
    def next(self, i: int) -> int:
        return (i + 1) % self.n
    
    def decrypt(self, code: List[int], k: int) -> List[int]:
        self.n = len(code)
        ans = [0] * self.n
        if not k:
            return ans
        l, r, s = 0, 0, 0
        if k > 0:
            l = 1
            for r in range(l, k + 1):
                s += code[r]
        else:
            r = self.n - 1
            for l in range(r, self.n + k - 1, -1):
                s += code[l]
        for i in range(self.n):
            ans[i] = s
            s -= code[l]
            l = self.next(l)
            r = self.next(r)
            s += code[r]
        return ans

File: Codes/1652-defuse-the-bomb_20240505.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-05-05 13:48:37
 * @LastEditors: LetMeFly
 * @LastEditTime: 2024-05-05 14:02:36
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class Solution {
private:
    int n;
    
    inline int next(int i) {
        return (i + 1) % n;
    }

public:
    vector<int> decrypt(vector<int>& code, int k) {
        n = code.size();
        vector<int> ans(n);
        if (!k) {
            return ans;
        }
        int l, r, s = 0;  // left, right, sum
        if (k > 0) {  // [1, k]
            l = 1;
            for (r = l; r <= k; r++) {
                s += code[r];
            }
            r--;
        }
        else {  // [n - (-k), n - 1]
            r = n - 1;
            for (l = r; l >= n + k; l--) {
                s += code[l];
            }
            l++;
        }
        for (int i = 0; i < n; i++) {
            ans[i] = s;
            s -= code[l];
            l = next(l);
            r = next(r);
            s += code[r];
            // printf("l = %d, r = %d, s = %d\n", l, r, s);  //***********
        }
        return ans;
    }
};

#ifdef _WIN32
/*
[5,7,1,4]
3

[12,10,16,13]
*/
int main() {
    string s;
    int k;
    while (cin >> s >> k) {
        vector<int> v = stringToVector(s);
        Solution sol;
        debug(sol.decrypt(v, k));
    }
    return 0;
}
#endif